File: 56.合并区间.java
import java.awt.List;
import java.util.ArrayList;

import jdk.internal.net.http.common.SequentialScheduler.RestartableTask;

/*
 * @lc app=leetcode.cn id=56 lang=java
 *
 * [56] 合并区间
 */

// @lc code=start
class Solution {
    public int[][] merge(int[][] intervals) {
        if(intervals.length == 0 || intervals == null)return new int[0][2];
        Arrays.sort(intervals, new Comparator<int[]>() {
            public int compare(int[] interval1, int[] interval2) {
                return interval1[0] - interval2[0];
            }
        });
        
        List<int[]> result = new ArrayList<>();
        for(int i = 0 ;i < intervals.length;i++){
            int left = intervals[i][0],right = intervals[i][1];
            if(result.size() == 0 || result.get(result.size() - 1)[1] < left){
                result.add(new int[]{left,right});
            }else{
                result.get(result.size() - 1)[1] = Math.max(result.get(result.size() - 1)[1] ,right);
            }
        }
        return result.toArray(new int[result.size()][]);
    }
}
// @lc code=end

File: 56.合并区间.cpp
/*
 * @lc app=leetcode.cn id=56 lang=cpp
 *
 * [56] 合并区间
 */

// @lc code=start
class Solution {
    static bool cmp(vector<int> &a, vector<int> &b){
        if(a[0] == b[0])return a[1] < b[1];
        return a[0] < b[0];
    }
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> result;
        if(intervals.size() == 0)return result;
        sort(intervals.begin(), intervals.end(), cmp);
        result.push_back(intervals[0]);
        for(int i = 1; i < intervals.size(); ++i){
            if(result.back()[1] >= intervals[i][0]){
                result.back()[1] = max(result.back()[1], intervals[i][1]);
            }else{
                result.push_back(intervals[i]);
            }
        }
        return result;
    }
};
// @lc code=end

